composer/checkpoint/state_dict.py

```diff
@@ -152,3 +155,110 @@ def _get_model_state_dict_with_fsdp_context_manager(model: nn.Module, sharded_st
     with FSDP.state_dict_type(model, state_dict_type=state_dict_type, state_dict_config=state_dict_config):
         model_state_dict = model.state_dict()
     return model_state_dict
+
+
+def get_metadata_state_dict(
+    model: Optional[Union[ComposerModel, nn.Module]] = None,
+    sharded_state_dict: Optional[bool] = None,
+    precision: Optional[Union[str, torch.dtype]] = None,
+    device: Optional[Device] = None,
+    device_train_microbatch_size: Optional[int] = None,
+    generate_parameter_info: Optional[bool] = False,
+) -> Dict[str, Any]:
+    """Generate the metadata and integrations for a training run.
+
+    Args:
+        model: The model to get the metadata state dict from. Only applicable if model is HuggingFaceModel
+        sharded_state_dict: Whether the checkpoint this metadata state dict is associated with is sharded or not.
+            Optional argument because this function may not be called in the context of making a full checkpoint.
+        precision: The precision of the model. Can be specified as a string ('fp32', 'fp16', 'bf16') or a torch.dtype.
+        device: The device the model is on.
+        device_train_microbatch_size: The microbatch size used for training on the device.
+        generate_parameter_info: Whether to generate parameter information for the model. Default is False.
+
+    This state dict includes:
+        * composer version
+        * composer commit hash
+        * gpu model
+        * num nodes
+        * num gpus
+        * num gpus per node
+        * cpu core count
+        * cpu model
+        * torch version
+        * python version
+        * optionally
+            * dist/communication backend
+            * precision
+            * hf model metadata
+            * device_train_microbatch_size
+            * sharded vs unsharded state dict
+            * model name
+            * param fqns, shapes, and requires_grad
+            * huggingface metadata
+
+    Returns:
+        The state dict containing the metadata and any integrations for a training run.
+    """
+
+    composer_env_dict = get_composer_env_dict()
+    ced = composer_env_dict
+
+    python_version = '.'.join([str(getattr(sys.version_info, k)) for k in ['major', 'minor', 'micro']])
+
+    metadata_state_dict = {
+        'composer_version': ced['composer_version'],
+        'composer_commit_hash': ced['composer_commit_hash'],
+        'torch_version': torch.__version__,
+        'python_version': python_version,
+        'num_nodes': ced['node_world_size'],
+        'num_gpus_per_node': ced['local_world_size'],
+        'num_gpus': dist.get_world_size(),
+        'gpu_model': ced['accelerator_model_name'],
+        'cpu_model': ced['host_processor_model_name'],
+        'cpu_core_count': ced['host_processor_core_count'],
```

Comment:
a lot of these are already saved in the metadata part of the checkpoint

Reply:
but @mvpatel2000 do we still need these?

Reply:
Some bugs are accelerator specific, so it's useful to have information. 

I'm not sure why we rename the key though, it should probably just be `accelerator_model_name` etc in the metadata as well. Can we keep it consistent in both places?

Reply:
seems like it would be easier to read this way?

Reply:
I'd prefer to keep consistent. Sometimes, it might not be a GPU for example so it's helpful to keep generic. Also just nice to use consistent names everywhere?